nomad/job_endpoint.go

```diff
@@ -414,3 +387,127 @@ func (j *Job) Evaluations(args *structs.JobSpecificRequest,
 	j.srv.setQueryMeta(&reply.QueryMeta)
 	return nil
 }
+
+// Plan is used to cause a dry-run evaluation of the Job and return the results
+// with a potential diff containing annotations.
+func (j *Job) Plan(args *structs.JobPlanRequest, reply *structs.JobPlanResponse) error {
+	if done, err := j.srv.forward("Job.Plan", args, args, reply); done {
+		return err
+	}
+	defer metrics.MeasureSince([]string{"nomad", "job", "plan"}, time.Now())
+
+	// Validate the arguments
+	if args.Job == nil {
+		return fmt.Errorf("Job required for plan")
+	}
+
+	// Initialize the job fields (sets defaults and any necessary init work).
+	args.Job.InitFields()
+
+	// Validate the job.
+	if err := validateJob(args.Job); err != nil {
+		return err
+	}
+
+	// Acquire a snapshot of the state
+	snap, err := j.srv.fsm.State().Snapshot()
+	if err != nil {
+		return err
+	}
+
+	// Get the original job
+	oldJob, err := snap.JobByID(args.Job.ID)
+	if err != nil {
+		return err
+	}
+
+	var index uint64
+	if oldJob != nil {
+		index = oldJob.JobModifyIndex + 1
+	}
+
+	// Insert the updated Job into the snapshot
+	snap.UpsertJob(index, args.Job)
+
+	// Create an eval and mark it as requiring annotations and insert that as well
+	eval := &structs.Evaluation{
+		ID:             structs.GenerateUUID(),
+		Priority:       args.Job.Priority,
+		Type:           args.Job.Type,
+		TriggeredBy:    structs.EvalTriggerJobRegister,
+		JobID:          args.Job.ID,
+		JobModifyIndex: index,
+		Status:         structs.EvalStatusPending,
+		AnnotatePlan:   true,
+	}
+
+	// Create an in-memory Planner that returns no errors and stores the
+	// submitted plan and created evals.
+	planner := scheduler.NewInMemoryPlanner()
+
+	// Create the scheduler and run it
+	sched, err := scheduler.NewScheduler(eval.Type, j.srv.logger, snap, planner)
+	if err != nil {
+		return err
+	}
+
+	if err := sched.Process(eval); err != nil {
+		return err
+	}
+
+	// Annotate and store the diff
+	annotations := planner.Plan.Annotations
+	if args.Diff {
+		jobDiff, err := oldJob.Diff(args.Job, true)
+		if err != nil {
+			return fmt.Errorf("failed to create job diff: %v", err)
+		}
+
+		if err := scheduler.Annotate(jobDiff, annotations); err != nil {
+			return fmt.Errorf("failed to annotate job diff: %v", err)
+		}
+		reply.Diff = jobDiff
+	}
+
+	reply.Cas = index
```

Comment:
I would just call it the `JobModifyIndex` instead of `CAS` since that is much clearer